### PR TITLE
Document MSC3758, MSC3966, MSC3873, and MSC3980.

### PR DESCRIPTION
**Pull request by clokep**
This documents the new push rule conditions:

* `event_property_is` ([MSC3758](https://github.com/matrix-org/matrix-spec-proposals/pull/3758))
* `event_property_contains` ([MSC3966](https://github.com/matrix-org/matrix-spec-proposals/pull/3966))

As well as the dotted-path escape rules from [MSC3873](https://github.com/matrix-org/matrix-spec-proposals/pull/3873) and [MSC3980](https://github.com/matrix-org/matrix-spec-proposals/pull/3980).

~~Based on #1461.~~

Preview: https://pr1464--matrix-spec-previews.netlify.app

### CODE REVIEW

File: data/api/client-server/definitions/sync_filter.yaml
Comment (clokep) [inline data/api/client-server/definitions/sync_filter.yaml:22]: The MSC doesn't *quite* say this since this implicitly adds escaping a backslash with another backslash, which I don't think was previously valid. I can back this out if you would like.

File: content/client-server-api/modules/push.md
Comment (clokep) [inline content/client-server-api/modules/push.md:313]: This isn't escaped properly:

```suggestion
For example, if `key` is `content.m\\.federate`, and `value` is `true`, then
```

I wanted to use an example of a real field -- if there are better ones than this, please let me know of ideas!

File: content/client-server-api/modules/push.md
Comment (anoadragon453) [reply content/client-server-api/modules/push.md:306]: Likewise for `event_property_contains`.

File: content/appendices.md
Comment (anoadragon453) [inline content/appendices.md:948]:
```suggestion
names with dots, e.g. `content.body` would represent a `body` property in the
```

File: content/appendices.md
Comment (anoadragon453) [inline content/appendices.md:957]:
```suggestion
other escape sequences are reserved for future use.
```

File: content/appendices.md
Comment (anoadragon453) [inline content/appendices.md:954]: I think it would help to also include an example of escaping a backslash as well (for a theoretical field, as none today contain backslashes).

File: changelogs/client_server/newsfragments/1464.clarification
Comment (anoadragon453) [inline changelogs/client_server/newsfragments/1464.clarification:1]:
```suggestion
Disambiguate using property names with dots in them during push rule processing, per [MSC3873](https://github.com/matrix-org/matrix-spec-proposals/pull/3873).
```

File: data/api/client-server/definitions/sync_filter.yaml
Comment (anoadragon453) [inline data/api/client-server/definitions/sync_filter.yaml:19]: This markdown link doesn't appear to work. I think you might need to include some more hyphens?

```suggestion
      fields are included. The entries are [dot-separated paths for each property](/appendices#dot-separated-property-paths)
```

Likewise for the other places the Appendices is referenced in this PR.

File: data/api/client-server/definitions/sync_filter.yaml
Comment (anoadragon453) [reply data/api/client-server/definitions/sync_filter.yaml:22]: Hmm, since MSC3873 doesn't propose modifying sync filters to support escaping backslashes (instead just taking the dot-escaping capability of sync filter), I believe the *correct* approach to unify the two would be to have another MSC about Sync filters to include support for escaping backslashes.

I think such an MSC would be met without any contention, as everyone just wants consistency at the end of the day.

However, while that separate MSC is in review and MSC3873 is merged, it does mean that clients/servers will have to match backslashes differently depending on whether they're processing push rules or sync filters, which is a bit of a pain. So this may be a case of the process getting in the way, and we simply edit MSC3873 after the fact to mention that sync filters should also escape backslashes.

Finally, this is all a bit moot as no fields in events currently contain backslashes anyway (except possibly some displaynames?).

What do other @matrix-org/spec-core-team members think?

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
This looks great, and I appreciate your clear wording.

A few bits and bobs to fix up, and a leading question.

File: content/client-server-api/modules/push.md
Comment (clokep) [reply content/client-server-api/modules/push.md:306]: The MSC discussed "non-compound values from Canonical JSON" (or something similar to that), is that phrasing useful or just list them out?

File: content/client-server-api/modules/push.md
Comment (clokep) [reply content/client-server-api/modules/push.md:313]: > Why don't we escape it properly here, since we're also introducing [MSC3873](https://github.com/matrix-org/matrix-spec-proposals/issues/3873)

Because I forgot to do it in the PR so was making a note to myself -- I was curious if a reviewer had a better property to check against.

File: data/api/client-server/definitions/sync_filter.yaml
Comment (clokep) [reply data/api/client-server/definitions/sync_filter.yaml:22]: > Finally, this is all a bit moot as no fields in events currently contain backslashes anyway (except possibly some displaynames?).

It only applies to property *names*, not values.

File: data/api/client-server/definitions/sync_filter.yaml
Comment (clokep) [reply data/api/client-server/definitions/sync_filter.yaml:19]: Oops. 👍

File: data/api/client-server/definitions/sync_filter.yaml
Comment (clokep) [reply data/api/client-server/definitions/sync_filter.yaml:22]: I decided to just write the MSC: see [MSC3980](https://github.com/matrix-org/matrix-spec-proposals/pull/3980).

Comment (anoadragon453) [reply data/api/client-server/definitions/sync_filter.yaml:22]: I'll try to push for review of the MSC on Tuesday with the rest of the Spec Core Team as it should be universally agreeable and a very short read. Thank you for writing it :slightly_smiling_face:

File: changelogs/client_server/newsfragments/1464.clarification
Comment (clokep) [inline changelogs/client_server/newsfragments/1464.clarification:1]:
```suggestion
Disambiguate using property names with dots in them during push rule processing, per [MSC3873](https://github.com/matrix-org/matrix-spec-proposals/pull/3873) and [MSC3980](https://github.com/matrix-org/matrix-spec-proposals/pull/3980).
```

File: data/api/client-server/definitions/sync_filter.yaml
Comment (clokep) [reply data/api/client-server/definitions/sync_filter.yaml:22]: @anoadragon453 Re-requested review, although this is blocked on the FCP timer of MSC3980.

Comment (clokep) [reply data/api/client-server/definitions/sync_filter.yaml:22]: This should be ready to go now. 👍

**Pull request merged**
